[Merged by Bors] - refactor(Algebra): replace `RingEquivClass.toRingEquiv` by structure-specific coercions

YaelDillies opened the pull request:
Remove the coercion from `E` to `RingEquiv` assuming `RingEquivClass E` and given by `RingEquivClass.toRingEquiv`. For each concrete type, reimplement this coercion through the relevant structure projection.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor commented on Mathlib/Algebra/Group/Equiv/Basic.lean:
```diff
@@ -29,15 +29,296 @@ end EmbeddingLike
+section coe
+
+@[to_additive]
+instance : EquivLike (M ≃* N) M N where
```
Isn't this section double with `Group/Equiv/Defs.lean`?

YaelDillies replied:
Very likely. I seem to have botched a rebase

Vierkantor commented on Mathlib/Algebra/Group/Equiv/Basic.lean:
```diff
@@ -29,15 +29,296 @@ end EmbeddingLike
+@[to_additive (attr := simp)]
+lemma toFun_eq_coe (f : M ≃* N) : f.toFun = f := rfl
+
+-- Porting note: `to_fun_eq_coe` no longer needed in Lean4
+
```
That porting note can go away, since we have `toFun_eq_coe` right above it :)
```suggestion
```

Vierkantor commented on Mathlib/Algebra/Order/Hom/Monoid.lean:
```diff
@@ -590,6 +590,32 @@ instance : Inhabited (α ≃*o α) :=
 
+/-- Symmetry of multiplication-preserving order isomorphisms -/
+@[to_additive (attr := symm) "Symmetry of addition-preserving order isomorphisms"]
+def symm (f : α ≃*o β) : β ≃*o α where
```
Surprising that we didn't have this before but it seems to indeed have been the case!

YaelDillies replied:
This ended up being added independently in #22402

Vierkantor commented on Mathlib/RingTheory/Valuation/Basic.lean:
```diff
@@ -236,18 +238,24 @@ theorem comap_comp {S₁ : Type*} {S₂ : Type*} [Ring S₁] [Ring S₂] (f : S
 
 /-- A `≤`-preserving group homomorphism `Γ₀ → Γ'₀` induces a map `Valuation R Γ₀ → Valuation R Γ'₀`.
 -/
-def map (f : Γ₀ →*₀ Γ'₀) (hf : Monotone f) (v : Valuation R Γ₀) : Valuation R Γ'₀ :=
```
Are these changes supposed to be included?

YaelDillies replied:
No indeed, I will fix when I get back to that PR (possibly in a while)

Vierkantor reviewed (commented):
Overall I agree that the generic hom coercions didn't turn out to be particularly useful, so I would be in favour of the PR.

Could you also check that the documentation on how to set up `FunLike`s is appropriately updated?

YaelDillies commented:
I am not touching this PR for a while by lack of time. Feel free to take over if you feel like it!

urkud commented:
Please add some explanations to the commit message.

YaelDillies commented:
The status is that it needs a lot more work before it's reviewable. I'm expecting the final diff to be around +-3k lines.

@j-loreaux suggested preliminary steps to this PR but I am not sure that one can actually take those steps before the whole PR.

Concretely, he suggests to first remove uses of hom classes in definitions, and then remove the generic coercions. The issue I see is that this will change the syntactic form of a lot of lemmas, and simp won't yet have been upgraded to handle that. My suggestion instead is to do the refactor one type of morphisms at a time, starting from the leaves of the morphism hierarchy and working towards the roots.

YaelDillies committed:
refactor(Algebra): replace `RingEquivClass.toRingEquiv` by structure-specific coercions
Remove the coercion from `E` to `RingEquiv` assuming `RingEquivClass E` and given by `RingEquivClass.toRingEquiv`.
For each concrete type, reimplement this coercion through the relevant structure projection.

artie2000 commented on Mathlib/RingTheory/AdjoinRoot.lean:
```diff
@@ -463,8 +463,8 @@ def mapAlgEquiv (f : S ≃ₐ[R] T) (p : S[X]) (q : T[X]) (h : Associated (p.map
     (mapAlgHom f p q h.symm.dvd)
     (mapAlgHom f.symm q p <| by
       -- FIXME: Coercion hell. See https://github.com/leanprover-community/mathlib4/pull/21031
```
needs updating?

Vierkantor replied:
Looks like the `RingHomClass.toRingHom` coercion needs to get the same treatment before we can get rid of this issue. If there already is an open PR for that, we should link it here (and otherwise turn it into a tracking issue?)

Vierkantor commented on Mathlib/Algebra/Algebra/Equiv.lean:
```diff
@@ -140,22 +140,21 @@ protected theorem coe_coe {F : Type*} [EquivLike F A₁ A₂] [AlgEquivClass F R
 
-instance hasCoeToRingEquiv : CoeOut (A₁ ≃ₐ[R] A₂) (A₁ ≃+* A₂) :=
-  ⟨AlgEquiv.toRingEquiv⟩
+instance : CoeOut (A₁ ≃ₐ[R] A₂) (A₁ ≃+* A₂) where coe := AlgEquiv.toRingEquiv
```
Since I noticed some discussion on the other PRs, let me explain that `CoeOut` is indeed the right way to go here: we have that the free variables on the left hand side **cannot** be inferred using the type on the right hand side, but the other way around they **can** be inferred.

`Coe` is what you would use in the opposite situation, when the variables on the left hand side **can** be inferred from the type on the right hand side (but not necessarily the opposite.

YaelDillies replied:
Thank you for the detailed explanation!

Vierkantor commented on Mathlib/Algebra/Order/Hom/Ring.lean:
```diff
@@ -316,6 +316,8 @@ instance : RingEquivClass (α ≃+*o β) α β where
   map_mul f := f.map_mul'
   map_add f := f.map_add'
 
+instance : CoeOut (α ≃+*o β) (α ≃+* β) where coe := toRingEquiv
```
This could either be a `Coe` or a `CoeOut`, based on the types. Since we generally drop (type) information moving from left to right in coercions like this, it's a good idea to standardize on `CoeOut` for this kind of coercion even if either way could work.

Vierkantor commented on Mathlib/RingTheory/RingInvo.lean:
```diff
@@ -79,6 +79,8 @@ instance : RingInvoClass (RingInvo R) R where
   map_mul f := f.map_mul'
   involution f := f.involution'
 
+instance : CoeTC (RingInvo R) (R ≃+* Rᵐᵒᵖ) where coe := toRingEquiv
```
`CoeTC` is an implementation detail and should not appear in new instances. From the types it looks like `Coe` or `CoeOut` could work, for consistency with above I vote for `CoeOut`.

```suggestion
instance : CoeOut (RingInvo R) (R ≃+* Rᵐᵒᵖ) where coe := toRingEquiv
```

(Unless, but I don't think this is the case, it causes a loop in which case we want `CoeHead`.)

Vierkantor reviewed (approved):
Thanks!

bors d+

YaelDillies committed:
update todo

YaelDillies committed:
CoeOut

YaelDillies commented:
bors merge
